<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [318]:
import psycopg2
import pandas as pd



In [319]:
connection = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    host='84.201.134.129',
    port=5432
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [320]:
# текст запроса
query_3_1 = f'''select count(*) 
from public.vacancies v
'''  #Запрос

df = pd.read_sql_query(query_3_1, connection)  #Получение результата
print(df) #Выведение результата



   count
0  49197


C:\Users\Admin\AppData\Local\Temp\ipykernel_13696\2165760488.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)  #Получение результата


In [321]:
# результат запроса
# Результат выведет значения количества вакансий 49197, что является нашим результатом

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [322]:
query_3_2 = f'''select count(*)
from public.employers 
''' #Запрос

df = pd.read_sql_query(query_3_2, connection)  #Получение результата
print(df) #Выведение результата


   count
0  23501


C:\Users\Admin\AppData\Local\Temp\ipykernel_13696\569136774.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)  #Получение результата


In [323]:
# текст запроса
# запрос, Получение результатаm, Выведение результата

In [324]:
# результат запроса
# Результат выведет значения количества работодателей 23501, что является нашим результатом

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [325]:
query_3_3 = f'''select count(*)
from public.areas 
''' #Запрос

df = pd.read_sql_query(query_3_3, connection)  #Получение результата
print(df) #Выведение результата

   count
0   1362


C:\Users\Admin\AppData\Local\Temp\ipykernel_13696\3479339141.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)  #Получение результата


In [326]:
# текст запроса
# запрос, Получение результатаm, Выведение результата

In [327]:
# результат запроса
# Результат выведет значения количества регионов 1362, что является нашим результатом

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [328]:
query_3_4 = f'''select count(*)
from public.industries 
''' #Запрос

df = pd.read_sql_query(query_3_4, connection)  #Получение результата
print(df) #Выведение результата

C:\Users\Admin\AppData\Local\Temp\ipykernel_13696\2326976954.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)  #Получение результата


   count
0    294


In [329]:
# текст запроса
# запрос, Получение результатаm, Выведение результата

In [330]:
# результат запроса
# Результат выведет значения количества сфер деятельности 294, что является нашим результатом

***

In [331]:
# выводы по предварительному анализу данных
# Результатом нашего пойска стали, получения общих количеств по таблицам поиска

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [332]:
query_4_1 = f'''SELECT a.name, COUNT(v.id)
FROM public.areas a
JOIN public.vacancies v ON a.id = v.area_id  
GROUP BY a.name
ORDER BY 2 DESC
'''  # Запрос

df = pd.read_sql_query(query_4_1, connection)  # Получение результата
print(df)  # Выведение результата


C:\Users\Admin\AppData\Local\Temp\ipykernel_13696\3978919220.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)  # Получение результата


                name  count
0             Москва   5333
1    Санкт-Петербург   2851
2              Минск   2112
3        Новосибирск   2006
4             Алматы   1892
..               ...    ...
764       Тарко-Сале      1
765    Новоаннинский      1
766            Бирск      1
767           Сасово      1
768             Анна      1

[769 rows x 2 columns]


In [333]:
# Фильтрация происходит с участием метода группировки(для каждого значения города, находится вакансия из списка вакансий, группируетсЯ по имени и выводит кол-во вакансий для каждого города)

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [334]:
query_4_2 = f'''SELECT count(id)
FROM public.vacancies
where salary_to is not null or salary_from is not null
'''  # Запрос

df = pd.read_sql_query(query_4_2, connection)  # Получение результата
print(df)  # Выведение результата

C:\Users\Admin\AppData\Local\Temp\ipykernel_13696\1971192768.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)  # Получение результата


   count
0  24073


In [335]:
# Проверяем на два условия, в одной из ячеик должен быть результат(Значения), кроме NULL. Используем условие Where 

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [336]:
query_4_3 = f'''SELECT ROUND(AVG(salary_to)) "Верхняя граница", ROUND(AVG(salary_from)) "Нижняя граница"
FROM public.vacancies

'''  # Запрос

df = pd.read_sql_query(query_4_3, connection)  # Получение результата
print(df)  # Выведение результата

   Верхняя граница  Нижняя граница
0         110537.0         71065.0


C:\Users\Admin\AppData\Local\Temp\ipykernel_13696\2714467818.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)  # Получение результата


In [337]:
# Используется двойной метод агрегации, сначала AVG потом ROUND

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [338]:
query_4_4 = f'''SELECT v.schedule, v.employment, count(*)
FROM public.vacancies v
group by 1, 2
order by 3 DESC
limit 1 offset 1
'''  # Запрос

df = pd.read_sql_query(query_4_4, connection)  # Получение результата
print(df)  # Выведение результата

           schedule        employment  count
0  Удаленная работа  Полная занятость   7802


C:\Users\Admin\AppData\Local\Temp\ipykernel_13696\3304728946.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)  # Получение результата


In [339]:
# Результат функции отсортированный список, предварительно сокращенный с помощью OFFSET и  LIMIT

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [340]:
query_4_5 = f'''SELECT  v.experience as "Требуемый опыт работы", COUNT(v.id) AS "Количество вакансий"
FROM public.vacancies v
GROUP BY 1
ORDER BY 2 ASC
'''  # Запрос

df = pd.read_sql_query(query_4_5, connection)  # Получение результата
print(df)  # Выведение результата

  Требуемый опыт работы  Количество вакансий
0           Более 6 лет                 1337
1             Нет опыта                 7197
2         От 3 до 6 лет                14511
3    От 1 года до 3 лет                26152


C:\Users\Admin\AppData\Local\Temp\ipykernel_13696\1182136083.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)  # Получение результата


In [341]:
# Результат: опыт работы в порядке возрастания вакансий

***

In [342]:
# Итогом данной работы стало поиск, определение, анализ вакансий на HH.ru. В работе представлены результаты исследования по представленным данным и их заданиям.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [343]:
query_5_1 = f'''SELECT e.name , COUNT(v.id) AS "Количество вакансий"
FROM public.vacancies v join public.employers e  on  v.employer_id = e.id -- JOIN 
GROUP BY 1
ORDER BY 2 DESC
limit 10
'''  # Запрос

df = pd.read_sql_query(query_5_1, connection)  # Получение результата
print(df)  # Выведение результата

                    name  Количество вакансий
0                 Яндекс                 1933
1             Ростелеком                  491
2               Тинькофф                  444
3                   СБЕР                  428
4          Газпром нефть                  331
5            ИК СИБИНТЕК                  327
6                    МТС                  292
7                DataArt                  247
8  Совкомбанк Технологии                  204
9             Первый Бит                  176


C:\Users\Admin\AppData\Local\Temp\ipykernel_13696\3924617457.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)  # Получение результата


In [344]:
# Результат, название компаний и количество вакансий предлагаемых компаний. ответ Яндекс, газпром нефть

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [ ]:
query_5_2 = f'''
SELECT --Запрос 
    a.name as region_name,
    COUNT(distinct e.id) as employer,
    COUNT(distinct v.id) as vacancy
    
FROM public.areas a -- Объединение через LEFT
LEFT JOIN public.employers e ON a.id = e.area
LEFT JOIN public.vacancies v ON a.id = v.area_id


where v.name is null
GROUP BY a.id, a.name

ORDER BY 2 DESc
'''  # Запрос

df = pd.read_sql_query(query_5_2, connection)  # Получение результата
print(df)  # Выведение результата

            region_name  employer  vacancy
0                Россия       410        0
1             Казахстан       207        0
2    Московская область        75        0
3    Краснодарский край        19        0
4    Ростовская область        18        0
..                  ...       ...      ...
588             Скидель         0        0
589          Белоозерск         0        0
590           Орловский         0        0
591          Павловская         0        0
592             Таксимо         0        0

[593 rows x 3 columns]


C:\Users\Admin\AppData\Local\Temp\ipykernel_13696\1576078651.py:18: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)  # Получение результата


In [346]:
# Результатом данного кода стала, таблица, в которой благодаря LEFT JOIN собрана информация о количестве вакансий и кол-ве работодателей по региону. Мы произвели фильтрацию по вакансиям, где 
# where v.name is null то есть количество вакансий равно(эквиваленто) = 0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [347]:
query_5_3 = f'''
SELECT
    e.name as employer_name,
    count(DISTINCT  v.area_id)
    
    
FROM public.employers e
join public.vacancies v on e.id = v.employer_id


GROUP BY
    e.id,
    e.name

ORDER BY 2 DESc
limit 15
'''  # Запрос

df = pd.read_sql_query(query_5_3, connection)  # Получение результата
print(df)  # Выведение результата

C:\Users\Admin\AppData\Local\Temp\ipykernel_13696\2931645569.py:19: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)  # Получение результата


                 employer_name  count
0                       Яндекс    181
1                   Ростелеком    152
2                   Спецремонт    116
3       Поляков Денис Иванович     88
4                    ООО ЕФИН      71
5                   Совкомбанк     63
6                          МТС     55
7   ЭФКО, Управляющая компания     49
8                 Почта России     48
9                         КРОН     48
10                       MCORE     46
11                 ИК СИБИНТЕК     46
12             Сбербанк-Сервис     45
13                       ANCOR     44
14                  Первый Бит     43


In [348]:
# Результатом будет таблица, в которой идет подсчет регионов в которых работодатель выставял свои вакансии ответ 181

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [349]:
query_5_4 = f''' SELECT COUNT(*) AS employers_without_industry
    
FROM public.employers e
left join public.employers_industries em on e.id = em.employer_id

WHERE
    em.industry_id IS NULL

ORDER BY 1 ASC
'''  # Запрос

df = pd.read_sql_query(query_5_4, connection)  # Получение результата
print(df)  # Выведение результата

C:\Users\Admin\AppData\Local\Temp\ipykernel_13696\1684674220.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)  # Получение результата


   employers_without_industry
0                        8419


In [350]:
# результат запроса. Кол-во работодателей у которых не указана свера деятельности. LEFT JOIN в данном запросе используется для збора всех данных включая пустых ячеик - деятельность  

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [351]:
query_5_5 = f''' select e.name, count(em.industry_id)
    
FROM public.employers e
join public.employers_industries em on e.id = em.employer_id


group by 1
HAVING COUNT(em.industry_id) = 4
ORDER BY 1 ASC
'''  # Запрос

df = pd.read_sql_query(query_5_5, connection)  # Получение результата
print(df)  # Выведение результата

                 name  count
0        101 Интернет      4
1            21vek.by      4
2                2ГИС      4
3                  2К      4
4         4 пикселя +      4
...               ...    ...
1131            ЮРИОН      4
1132  ЮТИП Технологии      4
1133           ЯКласс      4
1134            ЯрНео      4
1135    Яшин&Партнёры      4

[1136 rows x 2 columns]


C:\Users\Admin\AppData\Local\Temp\ipykernel_13696\2947285979.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)  # Получение результата


In [352]:
# результат запроса, таблица в которой собраны организации с предоставлением 4 типов сфер деятельномти

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [353]:
query_5_6 = f'''SELECT COUNT(DISTINCT e.id) AS employer_count
FROM public.employers e
JOIN public.employers_industries ei ON e.id = ei.employer_id
JOIN public.industries i ON ei.industry_id = i.id
WHERE i.name = 'Разработка программного обеспечения';
'''  # Запрос

df = pd.read_sql_query(query_5_6, connection) 
print(df)  # Выведение результата

C:\Users\Admin\AppData\Local\Temp\ipykernel_13696\1135561390.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


   employer_count
0            3553


In [354]:
# результат запроса,  используя несколько объектов loin, мы объединили табилцу employers_industries с public.employers, а так же industries с employers из чего следует фильтрация по профессии. Ответ 3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [355]:
import requests
from bs4 import BeautifulSoup

# Если таблица локально, поместите HTML в переменную или загрузите страницу с Wikipedia
# Для примера возьмём страницу с городами-миллионниками России
url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России'

response = requests.get(url)
response.raise_for_status()

soup = BeautifulSoup(response.text, 'html.parser')

# Находим все таблицы с классом, содержащим sortable (по примеру вашего класса)
tables = soup.find_all('table', class_='standard sortable jquery-tablesorter')

target_table = None
for table in tables:
    headers = [th.get_text(strip=True) for th in table.find_all('th')]
    if 'Город' in headers:
        target_table = table
        break

if not target_table:
    raise ValueError("Таблица с заголовком 'Город' не найдена")

# Нахождение индекса столбца "Город"
headers = [th.get_text(strip=True) for th in target_table.find_all('th')]
city_index = headers.index('Город')

# Извлечение городов из таблицы
cities = []
for row in target_table.find('tbody').find_all('tr'):
    cols = row.find_all('td')
    if len(cols) > city_index:
        city_cell = cols[city_index]
        # Обычно город в <a> теге внутри ячейки
        city_link = city_cell.find('a')
        city_name = city_link.get_text(strip=True) if city_link else city_cell.get_text(strip=True)
        cities.append(city_name)

# Вывод имен городов
print("Города-миллионники России:")
for city in cities:
    print(city)


ValueError: Таблица с заголовком 'Город' не найдена

In [ ]:
# результат запроса - не понял как выгружать, если покажите буду очень признателен  

***

In [ ]:
# выводы по анализу работодателей

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [ ]:
query_6_1 = f'''SELECT COUNT(id) 
FROM public.vacancies
WHERE name ILIKE '%data%' OR name ILIKE '%данн%'
'''  # Запрос

df = pd.read_sql_query(query_6_1, connection) 
print(df)  # Выведение результата

   count
0   1771


C:\Users\Admin\AppData\Local\Temp\ipykernel_15380\2953624814.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


In [ ]:
# результат запроса стало количество вакансий с наличием в запросе 'data' или 'данн'. Поиск по ключевым словам предоставляет возможность, быстро определить род деятельности человека 'data' - данные

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [ ]:
query_6_2 = '''
SELECT COUNT(id)
FROM public.vacancies
WHERE -- условия DS + key_skills
   (name iLIKE '%data scientist%'
   OR name iLIKE '%data science%'
   OR name iLIKE '%исследователь данных%'
   OR name ~* '\\mML\\M'  
   OR name iLIKE '%machine learning%'
   OR name iLIKE '%машинн%обучен%%')
AND ( name iLIKE '%junior%'  
   OR experience = 'Нет опыта'
   OR employment = 'Стажировка') 
''' # AND соединяется по двум наборам критериев так как приориттет AND стоит  выше чем у  OR

df = pd.read_sql_query(query_6_2, connection)
print(df)


C:\Users\Admin\AppData\Local\Temp\ipykernel_13696\1073711026.py:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


   count
0     51


In [ ]:
# результат запроса - количество вакансий = 51. Мы перебрали по показателям 

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [ ]:
query_6_3 = '''
SELECT COUNT(id)
FROM public.vacancies v
WHERE  -- условия DS + key_skills
(
    name ILIKE '%data scientist%' 
    OR name ILIKE '%data science%'
    OR name ILIKE '%исследователь данных%'
    OR (name ILIKE '%ML%' AND name NOT ILIKE '%HTML%')
    OR name ILIKE '%machine learning%'
    OR name ILIKE '%машинн%обучен%'
)
AND
(
    key_skills ILIKE '%SQL%' 
    OR key_skills ILIKE '%postgres%'
)
'''   # AND соединяется по двум наборам критериев так как приориттет AND стоит  выше чем у  OR

df = pd.read_sql_query(query_6_3, connection)
print(df)


C:\Users\Admin\AppData\Local\Temp\ipykernel_13696\2984017173.py:20: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


   count
0    229


In [ ]:
# результат запроса - количество вакансий = 229.  


4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [ ]:
query_6_4 = '''
SELECT COUNT(id)
FROM public.vacancies v
WHERE  -- условия DS
(
    name ILIKE '%data scientist%' 
    OR name ILIKE '%data science%'
    OR name ILIKE '%исследователь данных%'
    OR (name ILIKE '%ML%' AND name NOT ILIKE '%HTML%')
    OR name ILIKE '%machine learning%'
    OR name ILIKE '%машинн%обучен%'
)
AND
(
    key_skills ILIKE '%Python%' 
    
)
'''   # AND соединяется по двум наборам критериев так как приориттет AND стоит  выше чем у  OR

df = pd.read_sql_query(query_6_4, connection)
print(df)# текст запроса

   count
0    357


C:\Users\Admin\AppData\Local\Temp\ipykernel_13696\1467388676.py:20: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


In [ ]:
# результат запроса - количество вакансий = 357.  


5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [ ]:
query_6_5 = '''
SELECT 
round(AVG(CHAR_LENGTH(v.key_skills) - CHAR_LENGTH(replace(v.key_skills, CHR(9), ''))+1), 2) AS avg_skills_count

FROM public.vacancies v
WHERE -- условия DS
    (lower(name) LIKE '%data scientist%' OR
    lower(name) LIKE '%data science%' OR
    lower(name) LIKE '%исследователь данных%' OR
    (name LIKE '%ML%' AND lower(name) NOT LIKE'%html%') OR
    lower(name) LIKE '%machine learning%' OR
    lower(name) LIKE '%машинн%обучен%') AND
    (key_skills IS NOT NULL AND key_skills != '')
   
'''   # AND соединяется по двум наборам критериев так как приориттет AND стоит  выше чем у  OR

df = pd.read_sql_query(query_6_5, connection)
print(df)# текст запроса

C:\Users\Admin\AppData\Local\Temp\ipykernel_13696\1796127583.py:23: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


   avg_skills_count
0              6.41


In [ ]:
# результат запроса - кол-во mean ключевых вакансий = 6.41 

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [ ]:
query_6_6 = '''SELECT 
v.experience,
ROUND(AVG(CASE 
   WHEN   v.salary_to is not null and v.salary_from is not null THEN (v.salary_from + v.salary_to) / 2
   WHEN  v.salary_from is not null THEN v.salary_from     
   WHEN v.salary_to is not null THEN v.salary_to
   else null
   end
      ))  avg_salary
   
FROM public.vacancies v

WHERE -- условия DS
    (lower(name) LIKE '%data scientist%' OR
    lower(name) LIKE '%data science%' OR
    lower(name) LIKE '%исследователь данных%' OR
    (name LIKE '%ML%' AND lower(name) NOT LIKE'%html%') OR
    lower(name) LIKE '%machine learning%' OR
    lower(name) LIKE '%машинн%обучен%')
    
    AND ( name iLIKE '%junior%'  -- Запрос для новичков
   OR experience = 'Нет опыта'
   OR employment = 'Стажировка') 

GROUP BY v.experience
ORDER BY v.experience DESC  
'''   # AND соединяется по двум наборам критериев так как приориттет AND стоит  выше чем у  OR

df = pd.read_sql_query(query_6_6, connection)
print(df)# текст запроса

           experience  avg_salary
0  От 1 года до 3 лет     55000.0
1           Нет опыта     74643.0


C:\Users\Admin\AppData\Local\Temp\ipykernel_13696\4089439136.py:29: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


In [ ]:
# результат запроса -  У работников от 3 до 6 лет ЗП 243115. Но мне стало интересно проверить по запросу из задания 6.2 проверку на новичка и получается,
# что новичок без опыта работы получает больше чем с опытом работы от 1 года до 3 лет. 
# Это говорит нам, что на собеседовании нужно умалчивать об опыте работы Ж)

***

In [ ]:
# Модуль предлагает набор упражнений по разработке и исследованию запросов, внешне выглядящих стандартизированными и похожими друг на друга.
# Однако углубившись в суть каждого задания, можно обнаружить скрытую сложность даже там, где кажется всё простым.
# Это превращает процесс выполнения задач в увлекательное и полезное занятие для профессиональной сферы. Благодарю!

# Общий вывод по проекту

In [362]:
import plotly.express as px
import pandas as pd
from sqlalchemy import create_engine

user = 'skillfactory'
password = 'cCkxxLVrDE8EbvjueeMedPKt'
host = '84.201.134.129'
port = 5432
database = 'project_sql'

connection_string = f'postgresql://{user}:{password}@{host}:{port}/{database}'
engine = create_engine(connection_string)

sql_query = '''
SELECT
  a.name,
  COUNT(v.id) AS vacancies_count
FROM public.vacancies v
JOIN public.areas a ON v.area_id = a.id
GROUP BY 1
ORDER BY vacancies_count DESC;
'''

df = pd.read_sql(sql_query, engine)

# Создаем словарь с координатами по городам (названия в базе должны соответствовать)
coordinates = {
    'Москва': (55.7558, 37.6176),
    'Санкт-Петербург': (59.9343, 30.3351),
    'Минск': (53.9006, 27.5590),
    'Новосибирск': (55.0084, 82.9357),
    'Алматы': (43.2220, 76.8512),
    'Екатеринбург': (56.8389, 60.6057),
    'Нижний Новгород': (56.2965, 43.9361),
    'Казань': (55.8304, 49.0661),
    'Краснодар': (45.0355, 38.9753)
}

# Добавляем колонки с координатами на основе названия области/города
def get_lat(name):
    return coordinates.get(name, (None, None))[0]

def get_lon(name):
    return coordinates.get(name, (None, None))[1]

df['lat'] = df['name'].apply(get_lat)
df['lon'] = df['name'].apply(get_lon)

# Убираем строки без координат 
df = df.dropna(subset=['lat', 'lon'])

fig = px.scatter_mapbox(
    df,
    lat='lat',
    lon='lon',
    size='vacancies_count',
    hover_name='name',
    hover_data=['vacancies_count'],
    zoom=3,
    height=600,
    mapbox_style='open-street-map',
    title='Распределение вакансий по городам'
)

fig.show()

C:\Users\Admin\AppData\Local\Temp\ipykernel_13696\1194531715.py:52: DeprecationWarning: *scatter_mapbox* is deprecated! Use *scatter_map* instead. Learn more at: https://plotly.com/python/mapbox-to-maplibre/
  fig = px.scatter_mapbox(


График показывает, что большинство вакансий сосредоточено в Москве, что косвенно подтверждает высокую концентрацию предприятий в столичном регионе. Исходя из этого, можно сделать предположение, что существующий рынок труда формирует повышенный интерес среди специалистов, и в будущем примерно 5333 специалиста могут принять решение о переезде в Москву ради карьерных возможностей.